snuba/subscriptions/scheduler_processing_strategy.py

```diff
@@ -0,0 +1,150 @@
+        self.__max_ticks_buffered_per_partition = max_ticks_buffered_per_partition
+        self.__next_step = next_step
+        self.__metrics = metrics
+
+        self.__buffers: Mapping[int, Deque[Message[Tick]]] = {
+            index: deque() for index in range(self.__partitions)
+        }
+
+        # Stores the latest timestamp we received for any partition. This is
+        # just for recording the partition lag.
+        self.__latest_ts: Optional[datetime] = None
+
+    def poll(self) -> None:
+        pass
+
+    def submit(self, message: Message[Tick]) -> None:
+        # If the scheduler mode is immediate or there is only one partition
+        # immediately submit message to the next step.
+        # We don't keep any latest_ts values as it is not relevant.
+        if self.__mode == SchedulerMode.IMMEDIATE or self.__partitions == 1:
+            self.__next_step.submit(message)
+            return
+
+        # Update the latest_ts for metrics
+        if (
+            self.__latest_ts is None
+            or message.payload.timestamps.upper > self.__latest_ts
+        ):
+            self.__latest_ts = message.payload.timestamps.upper
+
+        tick_partition = message.payload.partition
+        self.__buffers[tick_partition].append(message)
+
+        # If the buffer length exceeds `max_ticks_buffered_per_partition`
+        # immediately submit the earliest message in that buffer to the next step.
+        if len(self.__buffers[tick_partition]) > cast(
+            int, self.__max_ticks_buffered_per_partition
+        ):
+            self.__next_step.submit(self.__buffers[tick_partition].popleft())
+            return
+
+        # If there are any empty buffers, we can't submit anything yet.
+        # Otherwise if all the buffers have ticks then we look for the partition/s
+        # with the earliest tick (i.e. the tick with the earliest upper timestamp
+        # interval value) and submit it to the next step.
+        if len(self.__buffers[tick_partition]) > 1:
+            return
+
+        while True:
+            earliest_ts = message.payload.timestamps.upper
+            earliest_ts_partitions = {message.payload.partition}
+
+            for partition_index in self.__buffers:
+                if partition_index == message.payload.partition:
+                    continue
+
+                buffer = self.__buffers[partition_index]
+                if len(buffer) == 0:
+                    return
+
+                tick = buffer[0].payload
+
+                partition_ts = tick.timestamps.upper
+
+                if partition_ts < earliest_ts:
+                    earliest_ts = tick.timestamps.upper
+                    earliest_ts_partitions = {tick.partition}
+
+                elif partition_ts == earliest_ts:
+                    earliest_ts_partitions.add(tick.partition)
+
+            for partition_index in earliest_ts_partitions:
+                self.__next_step.submit(self.__buffers[partition_index].popleft())
```

Comment:
So this means that you are producing all the same ticks you received in input, just you produce them when all partitions reached that timestamp. Which means you are producing twice the same tick (for two different partitions) if you had to ticks with the same `upper` timestamp.   What was the reason for going this way instead of producing a single watermark tick valid across partitions when that timestamp is reached ?